File: generate_sound.py
import math
import struct
import wave
import os
import sys

def generate_tone(frequency, duration, sample_rate=44100, volume=0.5):
    num_samples = int(sample_rate * duration)
    return [volume * math.sin(2 * math.pi * frequency * i / sample_rate) for i in range(num_samples)]

def apply_envelope(samples, attack_time=0.01, release_time=0.1, sample_rate=44100):
    attack_samples = int(sample_rate * attack_time)
    release_samples = int(sample_rate * release_time)
    total_samples = len(samples)
    
    env_samples = []
    for i, s in enumerate(samples):
        if i < attack_samples:
            env = i / attack_samples
        elif i > total_samples - release_samples:
            env = (total_samples - i) / release_samples
        else:
            env = 1.0
        env_samples.append(s * env)
    return env_samples

def save_wav(filename, samples, sample_rate=44100):
    with wave.open(filename, 'w') as wav_file:
        wav_file.setnchannels(1)
        wav_file.setsampwidth(2)
        wav_file.setframerate(sample_rate)
        for s in samples:
            val = int(s * 32767.0)
            val = max(-32768, min(32767, val))
            wav_file.writeframes(struct.pack('h', val))

output_path = sys.argv[1]
notes = [523.25, 659.25, 783.99, 1046.50]
all_samples = []
for i, freq in enumerate(notes):
    duration = 0.4 if i == len(notes) - 1 else 0.15
    samples = generate_tone(freq, duration, volume=0.4)
    samples = apply_envelope(samples, attack_time=0.02, release_time=duration-0.05)
    all_samples.extend(samples)

os.makedirs(os.path.dirname(output_path), exist_ok=True)
save_wav(output_path, all_samples)
print(f"Generated {output_path} successfully!")

File: generate_sound.js
const fs = require('fs');
const path = require('path');

function generateTone(frequency, duration, sampleRate = 44100, volume = 0.5) {
    const numSamples = Math.floor(sampleRate * duration);
    const samples = new Float32Array(numSamples);
    for (let i = 0; i < numSamples; i++) {
        samples[i] = volume * Math.sin(2 * Math.PI * frequency * i / sampleRate);
    }
    return samples;
}

function applyEnvelope(samples, attackTime = 0.01, releaseTime = 0.1, sampleRate = 44100) {
    const attackSamples = Math.floor(sampleRate * attackTime);
    const releaseSamples = Math.floor(sampleRate * releaseTime);
    const totalSamples = samples.length;
    
    const envSamples = new Float32Array(totalSamples);
    for (let i = 0; i < totalSamples; i++) {
        let env;
        if (i < attackSamples) {
            env = i / attackSamples;
        } else if (i > totalSamples - releaseSamples) {
            env = (totalSamples - i) / releaseSamples;
        } else {
            env = 1.0;
        }
        envSamples[i] = samples[i] * env;
    }
    return envSamples;
}

function saveWav(filename, allSamples, sampleRate = 44100) {
    const buffer = Buffer.alloc(44 + allSamples.length * 2);
    
    // RIFF chunk descriptor
    buffer.write('RIFF', 0);
    buffer.writeUInt32LE(36 + allSamples.length * 2, 4);
    buffer.write('WAVE', 8);
    
    // fmt sub-chunk
    buffer.write('fmt ', 12);
    buffer.writeUInt32LE(16, 16); // Subchunk1Size
    buffer.writeUInt16LE(1, 20); // AudioFormat
    buffer.writeUInt16LE(1, 22); // NumChannels
    buffer.writeUInt32LE(sampleRate, 24); // SampleRate
    buffer.writeUInt32LE(sampleRate * 2, 28); // ByteRate
    buffer.writeUInt16LE(2, 32); // BlockAlign
    buffer.writeUInt16LE(16, 34); // BitsPerSample
    
    // data sub-chunk
    buffer.write('data', 36);
    buffer.writeUInt32LE(allSamples.length * 2, 40);
    
    let offset = 44;
    for (let i = 0; i < allSamples.length; i++) {
        let val = Math.floor(allSamples[i] * 32767.0);
        val = Math.max(-32768, Math.min(32767, val));
        buffer.writeInt16LE(val, offset);
        offset += 2;
    }
    
    fs.writeFileSync(filename, buffer);
}

const outputPath = process.argv[2];
const notes = [523.25, 659.25, 783.99, 1046.50];
const allSamples = [];

for (let i = 0; i < notes.length; i++) {
    const freq = notes[i];
    const duration = (i === notes.length - 1) ? 0.4 : 0.15;
    let samples = generateTone(freq, duration, 44100, 0.4);
    samples = applyEnvelope(samples, 0.02, duration - 0.05, 44100);
    for (let j = 0; j < samples.length; j++) {
        allSamples.push(samples[j]);
    }
}

fs.mkdirSync(path.dirname(outputPath), { recursive: true });
saveWav(outputPath, allSamples);
console.log(`Generated ${outputPath} successfully!`);
